include/irr/asset/filters/CSwizzleAndConvertImageFilter.h

```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
 			#endif
 			auto perOutputRegion = [&blockDims,inFormat,outFormat,&state](const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
 			{
-				auto swizzle = [&commonExecuteData,&blockDims,inFormat,outFormat,&state](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos)
+				const auto inTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(inFormat);
+				const auto inChannelAmount = asset::getFormatChannelCount(inFormat);
+
+				// TESTING
+				auto getClampedSourcePixels = [&](auto sourcePixels)
+				{
+					constexpr auto MaxPlanes = 4;
+					auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(blockDims.y * blockDims.x * inTexelOrBlockByteSize);
+					const void* newSourcePixels[MaxPlanes] = { buffer->getPointer(), nullptr, nullptr, nullptr };
+
+					for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+						for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+						{
+							std::array<double, MaxPlanes> decodeBuffer = {};
+							asset::decodePixelsRuntime(inFormat, sourcePixels, decodeBuffer.data(), blockX, blockY);
+
+							for (auto& value : decodeBuffer)
+								core::clamp<double, double>(value, 0, 1);
+
+							asset::encodePixelsRuntime(inFormat, reinterpret_cast<uint8_t*>(buffer->getPointer()) + (blockY * blockDims.x + blockDims.x) * inTexelOrBlockByteSize, decodeBuffer.data());
+						}
+
+					return buffer;
+				};
+
+				auto swizzle = [&commonExecuteData,&blockDims,inFormat,outFormat,&state,&getClampedSourcePixels](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos)
 				{
 					constexpr auto MaxPlanes = 4;
 					const void* srcPix[MaxPlanes] = { commonExecuteData.inData+readBlockArrayOffset,nullptr,nullptr,nullptr };
 
+					auto clampedBuffer = getClampedSourcePixels(srcPix); // todo - compile time, do not waste memory
+					const void* newSourcePixels[MaxPlanes] = { clamp ? clampedBuffer->getPointer() : srcPix[0], nullptr, nullptr, nullptr }; // todo - compile time as well
+
 					for (auto blockY=0u; blockY<blockDims.y; blockY++)
 					for (auto blockX=0u; blockX<blockDims.x; blockX++)
 					{
+						std::array<double, MaxPlanes> decodeBuffer = {};
 						auto localOutPos = readBlockPos*blockDims+commonExecuteData.offsetDifference;
 						uint8_t* dstPix = commonExecuteData.outData+commonExecuteData.oit->getByteOffset(localOutPos,commonExecuteData.outByteStrides);
+
 						if constexpr(std::is_base_of<PolymorphicSwizzle,Swizzle>::value)
-							convertColor<Swizzle>(inFormat,outFormat,srcPix,dstPix,blockX,blockY,state->swizzle);
+							convertColor<Swizzle>(inFormat,outFormat, newSourcePixels,dstPix,blockX,blockY,state->swizzle);
 						else
-							convertColor<Swizzle>(inFormat,outFormat,srcPix,dstPix,blockX,blockY,*state);
+							convertColor<Swizzle>(inFormat,outFormat, newSourcePixels,dstPix,blockX,blockY,*state);
```

Comment:
instead of having that weird lambda at the start, you want to break down the `convertColor<>` functions into their implementation constituent `decodePixels` and `encodePixels` that use a `uint64_t[4]` (sometimes casted to a `double[4]`) buffer for temporary storage.

Then sometimes (when you have the template parameter) depending on a `if constexpr` you want to shove in a `clamp` between the decode and encode.

Reply:
because you have the silly lambda you feel the need to allocate temporary storage for the block

Reply:
all my comments basically boil down to:
- see how `convertColor<>` works
- decompose it into a separate decode+encode
- put a clamp between the decode and the encode depending on the constexpr bool template parameter

Reply:
adjusted in following commit